### PR TITLE
!ban

### PR DESCRIPTION
**Pull request by Padmanabh82**
#175 !ban command completed

### CODE REVIEW

File: src/features/commands.ts [hunk 1]
```diff
@@ -721,12 +721,36 @@ Instead:
...
+            {
+              title: "banned !!",
+              type: "rich",
+              description: `banned ${userid} for ${reason} successfully`,
+              color: EMBED_COLOR,
+            },
+          ],
+        })
+
+        const targetChannel = msg.guild?.channels.cache.get("938763995102470207") as TextChannel;
```
Comment (Padmanabh82) [inline src/features/commands.ts:742]: please add mod log channel id here

**Comment by nickmccurdy** [comment]
Please install the Prettier extension so you have the same format as us https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

File: src/features/commands.ts (hunk 1 above)
Comment (nickmccurdy) [reply src/features/commands.ts:742]: You should use `CHANNELS.modLog`

File: src/features/commands.ts [hunk 2]
```diff
@@ -638,47 +638,119 @@ Instead:
...
+        await msg.channel.send({
+          embeds: [
+            {
+              title: "tsk tsk",
+              type: "rich",
+              description: `${msg.author.toString()} This command is reserved for staff, Dont try to do this kind of shit`,
+              color: EMBED_COLOR,
+            },
+          ],
+        })
```
Comment (vcarl) [inline src/features/commands.ts:655]: Can you remove this and the `!unlock` changes? I appreciate the reasoning, but I think it's fine to just ignore commands from users who can't use them. It's good to keep PRs focused on 1 change at a time

File: src/features/commands.ts [hunk 3]
```diff
@@ -638,47 +638,119 @@ Instead:
...
+      } else {
```
Comment (vcarl) [inline src/features/commands.ts:692]: Here as well

File: src/features/commands.ts [hunk 4]
```diff
@@ -638,47 +638,119 @@ Instead:
...
-          guildTextChannel.guild.roles.everyone,
-          {
-            SEND_MESSAGES: null, // null will inherit the permission from the parent channel category
-          },
-        );
+    },
+  },
+  {
+    words: ["!ban"],
+    help: "used for banning users \n\t(usage: `!ban reason user`)",
```
Comment (vcarl) [inline src/features/commands.ts:708]:
```suggestion
    help: "Ban a member that can't be found from the Discord UI\n\t(usage: `!ban reason userId`)",
```

File: src/features/commands.ts [hunk 5]
```diff
@@ -638,47 +638,119 @@ Instead:
...
+        msg.channel.send({
+          embeds: [
+            {
+              title: "tsk tsk",
+              type: "rich",
+              description: `${msg.author.toString()} This command is reserved for staff, Dont try to do this kind of shit`,
+              color: EMBED_COLOR,
+            },
+          ],
+        })
```
Comment (vcarl) [inline src/features/commands.ts:722]: Same reasoning as above, silently failing for users who can't use it is fine. We also definitely do not want a message this rude -- "don't try to do this this kind of shit" is a pretty hostile way to say this
```suggestion
        return;
```

**Review by vcarl — CHANGES_REQUESTED** [review]

File: yarn.lock [hunk 6]
```diff
@@ -4,7 +4,7 @@
 
 "@discordjs/builders@^0.11.0":
   version "0.11.0"
-  resolved "https://registry.yarnpkg.com/@discordjs/builders/-/builders-0.11.0.tgz#4102abe3e0cd093501f3f71931df43eb92f5b0cc"
+  resolved "https://registry.npmjs.org/@discordjs/builders/-/builders-0.11.0.tgz"
```
Comment (vcarl) [inline yarn.lock:7]: Can you revert the changes to `yarn.lock` too? This looks like it might have been because of a version change or something, but I don't think it's necessary right now

File: src/features/commands.ts (hunk 5 above)
Comment (Padmanabh82) [reply src/features/commands.ts:722]: After reading this message no one will dare to run it again in a funny manner because its posted by bot

File: yarn.lock (hunk 6 above)
Comment (Padmanabh82) [reply yarn.lock:7]: I don't know anything about it. What should I do ????

File: src/features/commands.ts (hunk 4 above)
Comment (Padmanabh82) [reply src/features/commands.ts:708]: Ok i will

File: src/features/commands.ts (hunk 2 above)
Comment (Padmanabh82) [reply src/features/commands.ts:655]: Small change, so that's why i mixed with this

File: yarn.lock (hunk 6 above)
Comment (Padmanabh82) [reply yarn.lock:7]: @nickmccurdy whats your opinion ???

Comment (Padmanabh82) [reply yarn.lock:7]: I don't know anything about it, what should I do ???

File: src/features/commands.ts (hunk 5 above)
Comment (Padmanabh82) [reply src/features/commands.ts:722]: @nickmccurdy whats your opinion ???

Comment (ghardin137) [reply src/features/commands.ts:722]: It's even better if we don't allow users who aren't authorized to know this command exists.  It's information leakage

Comment (Padmanabh82) [reply src/features/commands.ts:722]: It won't affect how people will react to it, and if i will show to it publicly, people will get that if they left for some time the trick won't work

Comment (Padmanabh82) [reply src/features/commands.ts:722]: And how you would hide opensource code ???

Comment (BTMPL) [reply src/features/commands.ts:722]: The idea is that:

a) it just shouldn't react at all when a non-staff triggers the command
b) it's not polite/professional to use "this kind of shit" as a message

Comment (BTMPL) [reply src/features/commands.ts:722]: It doesn't matter if it's posted by a bot or not. It's unprofessional and only funny if you are 14 years old.

Comment (Padmanabh82) [reply src/features/commands.ts:722]: I am 11 (i think your refering my age)

Comment (Padmanabh82) [reply src/features/commands.ts:722]: Or let it be, any suggestions for saying same thing in polite manner ??

Comment (BTMPL) [reply src/features/commands.ts:722]: I'm referring to this being immature in general. 

IMO the command shouldn't output anything to non-mod users. Additionally, I would assume that mods would use it in a non-public channel anyway to avoid spam.

File: yarn.lock (hunk 6 above)
Comment (vcarl) [reply yarn.lock:7]: You should be able to do something like `git checkout main -- yarn.lock`, this has more info https://dev.to/lofiandcode/git-and-github-how-to-revert-a-single-file-dha

File: src/features/commands.ts (hunk 2 above)
Comment (vcarl) [reply src/features/commands.ts:655]: This isn't the behavior we want from `!lock` and `!unlock`, so we can't merge the PR until this and other requested changes have been made

File: .env.example [hunk 7]
```diff
@@ -1,4 +0,0 @@
-DISCORD_HASH=
-GITHUB_TOKEN=
-AMPLITUDE_KEY=
-BOT_LOG=
```
Comment (nickmccurdy) [inline .env.example:4]: We still need this file to help contributors set up the bot locally

File: src/features/commands.ts [hunk 8]
```diff
@@ -638,47 +638,119 @@ Instead:
...
-          },
-        );
+    },
+  },
+  {
+    words: ["!ban"],
+    help: "used for banning users \n\t(usage: `!ban reason user`)",
+    category: "Moderation",
+    handleMessage: (msg) => {
+      if (!msg.guild || !isStaff(msg.member)) {
```
Comment (nickmccurdy) [inline src/features/commands.ts:711]: **Note:** This would also allows MVPs to ban, which is different from our Discord permissions

**Review by nickmccurdy — CHANGES_REQUESTED** [review]

File: src/features/commands.ts [hunk 9]
```diff
@@ -638,47 +638,119 @@ Instead:
...
+            {
+              title: "tsk tsk",
+              type: "rich",
+              description: `${msg.author.toString()} This command is reserved for staff, Dont try to do this kind of shit`,
+              color: EMBED_COLOR,
+            },
+          ],
+        })
+      } else {
+        const [command, reason, userid] = msg.content.split(" ")
```
Comment (nickmccurdy) [inline src/features/commands.ts:724]: `command` needs to be omitted since it's not used

**Review by nickmccurdy — CHANGES_REQUESTED** [review]

File: src/features/commands.ts (hunk 8 above)
Comment (Padmanabh82) [reply src/features/commands.ts:711]: So, is it ok ??? Or should I change behaviour of it ??

File: src/features/commands.ts (hunk 9 above)
Comment (Padmanabh82) [reply src/features/commands.ts:724]: Ya it's not in use but in array the command also comes. So, fro work all things i need to assign it some where

Comment (Padmanabh82) [reply src/features/commands.ts:724]: Ok i will try to remove it

File: .env.example (hunk 7 above)
Comment (Padmanabh82) [reply .env.example:4]: Why ?? You could just do same with real .env file. Just remove keys simple

File: src/features/commands.ts (hunk 5 above)
Comment (Padmanabh82) [reply src/features/commands.ts:722]: Ok, i will remove it

File: src/features/commands.ts (hunk 1 above)
Comment (Padmanabh82) [reply src/features/commands.ts:742]: ok, thanks

File: src/features/commands.ts (hunk 9 above)
Comment (Padmanabh82) [reply src/features/commands.ts]: done !

File: src/features/commands.ts (hunk 4 above)
Comment (Padmanabh82) [reply src/features/commands.ts]: done !

File: yarn.lock (hunk 6 above)
Comment (Padmanabh82) [reply yarn.lock:7]: done !

File: src/features/commands.ts (hunk 8 above)
Comment (Padmanabh82) [reply src/features/commands.ts:726]: @vcarl and @nickmccurdy what should i do ?? or it is ok ??

**Comment by Padmanabh82** [comment]
@vcarl @nickmccurdy see i have made new commits

File: src/features/commands.ts [hunk 10]
```diff
@@ -4,10 +4,45 @@ import { Message, TextChannel } from "discord.js";
 import cooldown from "./cooldown";
 import { ChannelHandlers } from "../types";
 import { isStaff } from "../helpers/discord";
+import { CHANNELS } from "../constants";
 
 export const EMBED_COLOR = 7506394;
 
-type Categories = "Reactiflux" | "Communication" | "Web" | "React/Redux";
+const Emojies = [
```
Comment (Padmanabh82) [inline src/features/commands.ts:11]: A ester egg for fun 😊

File: src/features/commands.ts [hunk 11]
```diff
@@ -638,47 +674,81 @@ Instead:
...
       }
+    },
+  },
+  {
+    words: ["!ban"],
+    help: "Ban a member that can't be found from the Discord UI\n\t(usage: `!ban reason userId`)",
+    category: "Moderation",
+    handleMessage: (msg) => {
+      if (!msg.guild || !isStaff(msg.member)) {
+        msg.react(`<${Emojies[Math.floor(Math.random() * Emojies.length)]}>`);
```
Comment (Padmanabh82) [inline src/features/commands.ts:727]: Used ester egg here

File: src/features/commands.ts [hunk 12]
```diff
@@ -672,48 +679,74 @@ Instead:
...
+            guildTextChannel.guild.roles.everyone,
+            {
+              SEND_MESSAGES: false,
+              CREATE_PUBLIC_THREADS: false,
+              CREATE_PRIVATE_THREADS: false,
+              SEND_MESSAGES_IN_THREADS: false,
+            },
+          );
+          guildTextChannel.send("This channel has been locked by a moderator");
+        }
```
Comment (vcarl) [inline src/features/commands.ts:699]: Please revert this entire change, since functionality isn't affected we shouldn't change the code

File: src/features/commands.ts [hunk 13]
```diff
@@ -672,48 +679,74 @@ Instead:
...
+          const { channel: guildTextChannel } = msg;
+          await guildTextChannel.permissionOverwrites.create(
+            guildTextChannel.guild.roles.everyone,
+            {
+              SEND_MESSAGES: null, // null will inherit the permission from the parent channel category
+            },
+          );
+        }
+    },
+  },
```
Comment (vcarl) [inline src/features/commands.ts:719]: Please revert this entire change

**Review by vcarl — CHANGES_REQUESTED** [review]

**Comment by Padmanabh82** [comment]
Done !

File: src/features/commands.ts (hunk 13 above)
Comment (Padmanabh82) [reply src/features/commands.ts]: Done

File: src/features/commands.ts (hunk 12 above)
Comment (Padmanabh82) [reply src/features/commands.ts]: Done !

**Comment by nickmccurdy** [comment]
Overall the changes look good, but ESLint is still failing on some formatting changes. You can reformat the file using Prettier (either via the extension or `yarn prettier`) to see if it fixes it. Then make sure `yarn test` passes locally, and then push commits again.

**Review by nickmccurdy — CHANGES_REQUESTED** [review]
Overall the changes look good, but ESLint is still failing on some formatting changes:
1. Reformat the file using Prettier (either via the extension or `yarn prettier`) to see if it fixes it
2. Make sure `yarn test` passes locally
3. Commit and push

**Comment by Padmanabh82** [comment]
@nickmccurdy now ???

**Comment by vcarl** [comment]
This PR had been in a mergable state, but now that you've pushed entirely new functionality, we'll need to re-review it

**Comment by Padmanabh82** [comment]
Done !!

**Comment by Padmanabh82** [comment]
Oops !! forgot to run tests
